src/pubtools/sign/clients/registry.py

```diff
@@ -0,0 +1,168 @@
+        """
+        parsed = urlparse(image_reference)
+        if not parsed.scheme:
+            parsed = urlparse(f"docker://{image_reference}")
+        registry = parsed.netloc
+        existing_auth_files = []
+        auth_files = AUTH_FILES if not self.auth_file else [self.auth_file] + AUTH_FILES
+        for af in auth_files:
+            if os.path.exists(os.path.expandvars(af)):
+                existing_auth_files.append(os.path.expandvars(af))
+        for eaf in existing_auth_files:
+            parsed_conf = json.load(open(os.path.expandvars(eaf)))
+            parsed_auths = parsed_conf.get("auths")
+            if registry in parsed_auths:
+                auth = (
+                    base64.b64decode(parsed_auths.get(registry)["auth"].encode("utf-8"))
+                    .decode("utf-8")
+                    .split(":")
+                )
+                break
+        else:
+            auth = (self.username, self.password)
+        return auth
+
+    def authenticate_to_registry(self, image_reference: str, auth_header: str) -> str:
+        """Ask for auth token based on given auth header.
+
+        Ask for authentication token based on the given auth header.
+
+        Args:
+            image_reference (str): Image reference to resolve authentication for.
+            auth_header (str): Authentication header from the registry.
+        Returns:
+            str: Authentication token.
+        """
+        _, _, value = auth_header.partition("Bearer")
+        items = parse_http_list(value)
+        opts = parse_keqv_list(items)
+        unparse_parts = [
+            "https",
+            opts["realm"].replace("https://", "").split("/", 1)[0],
+            opts["realm"].replace("https://", "").split("/", 1)[1],
+            "",
+            urlencode({"service": opts["service"], "scope": opts["scope"]}),
+            "",
+        ]
+        auth_url = urlunparse(unparse_parts)
+        username, password = self.resolve_authentication(image_reference)
+        response = self.session.get(auth_url, auth=(username, password))
+        if response.status_code == 200:
+            return response.json().get("token")
+        else:
+            response.raise_for_status()
+
+    def check_container_image_exists(
+        self, image_reference: str, auth_token: Optional[AuthTokenWrapper] = None
+    ) -> bool:
+        """Check if the given container image exists.
+
+        Args:
+            image_reference (str): Image reference to check.
+            auth_token (Optional[AuthTokenWrapper]): Authentication token.
+        Returns:
+            bool: True if the image exists, False otherwise.
+        """
+        repo_ref, tag = image_reference.rsplit(":", 1)
+        registry, repo = repo_ref.split("/", 1)
+        manifest_url = f"https://{registry}/v2/{repo}/manifests/{tag}"
+        headers = {"Authorization": f"Bearer {auth_token.token}"}
+        response = self.session.get(manifest_url, headers=headers)
```

Comment:
If more methods interacting with registry are added to this class, will all of them contain the authentication code?

Reply:
yeah and in that case it would be worth of extra method, but so far I don't see any other use cases to check registry for anything.